pkg/proxy/runner/bounded_frequency_runner_test.go

```diff
@@ -17,437 +17,474 @@ limitations under the License.
 package runner
 
 import (
-	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
+
+	clock "k8s.io/utils/clock/testing"
 )
 
 // Track calls to the managed function.
 type receiver struct {
-	lock    sync.Mutex
-	run     bool
-	retryFn func()
+	counter atomic.Int32
 }
 
 func (r *receiver) F() {
-	r.lock.Lock()
-	defer r.lock.Unlock()
-	r.run = true
-
-	if r.retryFn != nil {
-		r.retryFn()
-		r.retryFn = nil
-	}
+	r.counter.Add(1)
 }
 
-func (r *receiver) reset() bool {
-	r.lock.Lock()
-	defer r.lock.Unlock()
-	was := r.run
-	r.run = false
-	return was
+func (r *receiver) calls() int {
+	return int(r.counter.Load())
 }
 
-func (r *receiver) setRetryFn(retryFn func()) {
-	r.lock.Lock()
-	defer r.lock.Unlock()
-	r.retryFn = retryFn
+func (r *receiver) waitForValue(n int) bool {
+	for i := 0; i < 10; i++ {
+		if r.calls() == n {
+			return true
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
+	return false
 }
 
-// A single change event in the fake timer.
-type timerUpdate struct {
-	active bool
-	next   time.Duration // iff active == true
-}
+func Test_BoundedFrequencyRunner(t *testing.T) {
+	var minInterval = 1 * time.Second
+	var retryInterval = 5 * time.Second
+	var maxInterval = 10 * time.Second
+	obj := &receiver{}
+	fakeClock := clock.NewFakeClock(time.Now())
+	runner := construct("test-runner", obj.F, minInterval, retryInterval, maxInterval, fakeClock)
+	stop := make(chan struct{})
+	defer close(stop)
 
-// Fake time.
-type fakeTimer struct {
-	c chan time.Time
+	go runner.Loop(stop)
+	// Wait for the loop to be running.
+	time.Sleep(20 * time.Millisecond)
 
-	lock    sync.Mutex
-	now     time.Time
-	timeout time.Time
-	active  bool
+	// Run once, immediately.
+	// rel=0ms
+	runner.Run()
+	if !obj.waitForValue(1) {
+		t.Errorf("expected one call, got %d", obj.calls())
+	}
 
-	updated chan timerUpdate
-}
+	// Run again, before minInterval expires.
+	fakeClock.Step(500 * time.Millisecond) // rel=500ms
+	runner.Run()
+	if !obj.waitForValue(1) {
+		t.Errorf("expected one call, got %d", obj.calls())
+	}
```

Comment:
The name "waitForValue" is confusing here since what you really mean is "wait a bit and confirm that the value didn't change".

I don't like the way this unit test works. I feel like it would be better to change how the runner works internally so that the unit test can be completely race-free.

Reply:
the problem is that I reused the helper and the name is bad, for fakeclock test I should not use this, as you don't need to active wait for the call since we control the clock steps ... I was trying to keep the diff small but I make it more confusing